File: Post Requests/post_example.py
# Use these functions to make an asynchronous POST request to the server

import requests
import asyncio
# You must install ctime_binding to have this method in your python path
from ctime_binding import get_c_time

# Send a post request
def log_time(code):
    asyncio.run(_log_time(code))

# async helper methodd
async def _log_time(code):
    # Get the time from the c binding
    now = get_c_time()
    sec = now[0]
    usec = now[1]

    # Send a post request
    url = "http://localhost:4560/report"

    time_stamp = {
        "identifier": code,
        "time": {
            "sec": sec,
            "usec": usec
        }
    }
    print("LOGGING IN PYTHON: ", time_stamp)

    requests.post(url, json = time_stamp)

File: Post Requests/post_example.c
# Use these functions to make an asynchronous POST request to the server

#include <curl/curl.h>
#include <pthread.h>
#include <syslog.h>
#include <time.h>


// Helper function to run on separate thread
void *_log_time(void *code) {
	CURL *curl;
	CURLcode res;

	// Get the code
	char *identifier = (char *)code;

	// Get the time
	struct timeval now;
	gettimeofday(&now, NULL);

	// Create the json object
	char *unfilledJson =
			"{ \"identifier\": \"%s\", \"time\": { \"sec\": \"%ld\", \"usec\": \"%ld\" } }";
	char jsonObjString[300];
	sprintf(jsonObjString, unfilledJson, identifier, now.tv_sec, now.tv_usec);

	curl = curl_easy_init();
	if (curl) {
		// Fill the headers
		struct curl_slist *headers = NULL;
		headers = curl_slist_append(headers, "Accept: application/json");
		headers = curl_slist_append(headers, "Content-Type: application/json");
		headers = curl_slist_append(headers, "charset: utf-8");

		curl_easy_setopt(curl, CURLOPT_URL, "http://localhost:4560/report");
		// Specify the POST data
		curl_easy_setopt(curl, CURLOPT_CUSTOMREQUEST, "POST");
		curl_easy_setopt(curl, CURLOPT_HTTPHEADER, headers);
		curl_easy_setopt(curl, CURLOPT_POSTFIELDS, jsonObjString);

		// Perform the request
		res = curl_easy_perform(curl);
		if (res != CURLE_OK) {
			fprintf(stderr, "curl_easy_perform() failed: %s\n", curl_easy_strerror(res));
		}

		// Cleanup
		curl_easy_cleanup(curl);
	}

	// Finish the separate processs
	return NULL;
}

// Call this function with a 3 letter code representing the type of timestamp
void log_time(char *code) {
	// Declare a var for the id
	pthread_t id;

	pthread_create(&id, NULL, _log_time, code);
}

